src/grad.jl

```diff
@@ -35,15 +35,66 @@ end
 jacobian(fdm, f, x::Vector{<:Real}) = jacobian(fdm, f, x, length(f(x)))
 
 """
-    jvp(fdm, f, x::AbstractVector{<:Real}, ẋ::AbstractVector{<:Real})
+    _jvp(fdm, f, x::Vector{<:Real}, ẋ::AbstractVector{<:Real})
 
 Convenience function to compute `jacobian(f, x) * ẋ`.
 """
-jvp(fdm, f, x::Vector{<:Real}, ẋ::AV{<:Real}) = jacobian(fdm, f, x) * ẋ
+_jvp(fdm, f, x::Vector{<:Real}, ẋ::AV{<:Real}) = jacobian(fdm, f, x) * ẋ
 
 """
-    j′vp(fdm, f, ȳ::AbstractVector{<:Real}, x::AbstractVector{<:Real})
+    _j′vp(fdm, f, ȳ::AbstractVector{<:Real}, x::Vector{<:Real})
 
 Convenience function to compute `jacobian(f, x)' * ȳ`.
 """
-j′vp(fdm, f, ȳ::AV{<:Real}, x::Vector{<:Real}) = jacobian(fdm, f, x, length(ȳ))' * ȳ
+_j′vp(fdm, f, ȳ::AV{<:Real}, x::Vector{<:Real}) = jacobian(fdm, f, x, length(ȳ))' * ȳ
+
+"""
+    jvp(fdm, f, x, ẋ)
+
+Compute a Jacobian-vector product with any types of arguments for which `to_vec` is defined.
+"""
+function jvp(fdm, f, (x, ẋ)::Tuple{Any, Any})
+    x_vec, vec_to_x = to_vec(x)
+    _, vec_to_y = to_vec(f(x))
+    return vec_to_y(_jvp(fdm, x_vec->to_vec(f(vec_to_x(x_vec)))[1], x_vec, to_vec(ẋ)[1]))
+end
+function jvp(fdm, f, xẋs::Tuple{Any, Any}...)
+    x, ẋ = collect(zip(xẋs...))
+    return jvp(fdm, xs->f(xs...), (x, ẋ))
+end
+
+"""
+    j′vp(fdm, f, ȳ, x...)
+
+Compute an adjoint with any types of arguments for which `to_vec` is defined.
+"""
+function j′vp(fdm, f, ȳ, x)
+    x_vec, vec_to_x = to_vec(x)
+    ȳ_vec, _ = to_vec(ȳ)
+    return vec_to_x(_j′vp(fdm, x_vec->to_vec(f(vec_to_x(x_vec)))[1], ȳ_vec, x_vec))
+end
+j′vp(fdm, f, ȳ, xs...) = j′vp(fdm, xs->f(xs...), ȳ, xs)
+
+# Transform `x` into a vector, and return a closure which inverts the transformation.
+to_vec(x::Real) = ([x], x->x[1])
```

Comment:
```suggestion
to_vec(x::Real) = ([x], first)
```
Not a big deal or anything, just seems nice to be able to avoid the anonymous function.